lib/ansible/modules/packaging/os/dnf.py

```diff
@@ -724,38 +724,31 @@ def _mark_package_install(self, pkg_spec, upgrade=False):
         is_newer_version_installed = self._is_newer_version_installed(pkg_spec)
         is_installed = self._is_installed(pkg_spec)
         try:
-            if self.allow_downgrade:
-                # dnf only does allow_downgrade, we have to handle this ourselves
-                # because it allows a possibility for non-idempotent transactions
-                # on a system's package set (pending the yum repo has many old
-                # NVRs indexed)
-                if upgrade:
-                    if is_installed:
-                        self.base.upgrade(pkg_spec)
+            if is_newer_version_installed:
+                if self.allow_downgrade:
+                    # TODO: Copied from existing code, no idea how this really works
+                    # dnf only does allow_downgrade, we have to handle this ourselves
+                    # because it allows a possibility for non-idempotent transactions
+                    # on a system's package set (pending the yum repo has many old
+                    # NVRs indexed)
+                    if upgrade:
+                        if is_installed:
+                            self.base.upgrade(pkg_spec)
+                        else:
+                            self.base.install(pkg_spec)
                     else:
                         self.base.install(pkg_spec)
-                else:
-                    self.base.install(pkg_spec)
-            elif not self.allow_downgrade and is_newer_version_installed:
```

Comment:
Where is the `allow_downgrade` scenario being addressed in this code change?

Reply:
As far as I understood it downgrade only ever makes sense if a newer version is installed, so it is in:
https://github.com/ansible/ansible/pull/66209/files#diff-87843c1a75e8a1dc7759ae78dcadf890R727-R728

of my changes. All the other sections shouldn't have to bother about downgrades at all because if the version installed is not newer it is either an upgrade or a reinstall, but how would this be a downgrade?

Reply:
Note: I completely could miss some essential parts of how `allow_downgrade` is supposed to work, but my changes do not break the tests, so we could have missing coverage here.

An example of a (new) breaking test would be much appreciated if you think that there is a flow in my control flow.

Reply:
Nevermind me, I follow. Your control flow is better :+1: